hydra-node/test/Hydra/Chain/Direct/TxSpec.hs

```diff
@@ -141,6 +187,215 @@ spec =
+
+addRequiredSignatures :: [VerificationKey PaymentKey] -> Tx -> Tx
+addRequiredSignatures vks tx =
+  fromLedgerTx $
+    toLedgerTx tx
+      & bodyTxL . reqSignerHashesTxBodyL .~ Set.fromList (toLedgerKeyHash . verificationKeyHash <$> vks)
+
+genBlueprintTxWithUTxO :: Gen (UTxO, Tx)
+genBlueprintTxWithUTxO =
+  fmap (second unsafeBuildTransaction) $
+    spendingPubKeyOutput (mempty, emptyTxBody)
+      >>= spendSomeScriptInputs
+      >>= addSomeReferenceInputs
+      >>= addValidityRange
+      >>= addRandomMetadata
+      >>= removeRandomInputs
+      >>= addCollateralInputs
+ where
+  spendingPubKeyOutput (utxo, txbody) = do
+    utxoToSpend <- (genUTxOAdaOnlyOfSize . getPositive) . Positive =<< choose (0, 50)
+    pure
+      ( utxo <> utxoToSpend
+      , txbody & addVkInputs (toList $ UTxO.inputSet utxoToSpend)
+      )
+
+  spendSomeScriptInputs (utxo, txbody) = do
+    let alwaysSucceedingScript = PlutusScriptSerialised $ Plutus.alwaysSucceedingNAryFunction 3
+    n <- choose (1, 100)
+    datum <- unsafeHashableScriptData . fromPlutusData . I <$> arbitrary
+    redeemer <-
+      unsafeHashableScriptData . fromPlutusData . B . BS.pack
+        <$> vector n
+    let genTxOut = do
+          value <- balance @Tx <$> genUTxOAdaOnlyOfSize n
+          tokens <- genSomeTokens
+          let scriptAddress = mkScriptAddress testNetworkId alwaysSucceedingScript
+          pure $ TxOut scriptAddress (value <> tokens) (TxOutDatumInline datum) ReferenceScriptNone
+    utxoToSpend <- genUTxO1 genTxOut
+    pure
+      ( utxo <> utxoToSpend
+      , txbody
+          & addInputs
+            ( UTxO.pairs $
+                ( \_ ->
+                    BuildTxWith $
+                      ScriptWitness ScriptWitnessForSpending $
+                        mkScriptWitness alwaysSucceedingScript (ScriptDatumForTxIn datum) redeemer
+                )
+                  <$> utxoToSpend
+            )
+      )
+
+  addSomeReferenceInputs (utxo, txbody) = do
+    txout <- genTxOutWithReferenceScript
+    txin <- arbitrary
+    pure (utxo <> UTxO.singleton (txin, txout), txbody & addReferenceInputs [txin])
+
+  addValidityRange (utxo, txbody) = do
+    (start, end) <- arbitrary
+    pure
+      ( utxo
+      , txbody{txValidityLowerBound = start, txValidityUpperBound = end}
+      )
+
+  addRandomMetadata (utxo, txbody) = do
+    mtdt <- oneof [randomMetadata, pure TxMetadataNone]
+    pure (utxo, txbody{txMetadata = mtdt})
+   where
+    randomMetadata = do
+      n <- choose (2, 50)
+      metadata <- Text.take n <$> genSomeText
+      l <- arbitrary
+      pure $
+        TxMetadataInEra $
+          TxMetadata $
+            Map.fromList [(l, TxMetaText metadata)]
+
+  removeRandomInputs (utxo, txbody) = do
+    someInput <- elements $ txIns txbody
+    pure (utxo, txbody{txIns = [someInput]})
+
+  addCollateralInputs (utxo, txbody) = do
+    utxoToSpend <- (genUTxOAdaOnlyOfSize . getPositive) . Positive =<< choose (0, 50)
```

Comment:
Multiple collateral inputs is quite uncommon (and frankly I don't know why one would want that). Testing with no collateral and one collateral input is enough.